src/util/psbt/roles.rs

```diff
@@ -0,0 +1,373 @@
+    /// (for P2PK and P2WPK outputs)
+    fn add_input_derivation(&mut self, index: u32, pubkey: &PublicKey, fingerprint: Fingerprint, path: &DerivationPath) -> Result<&mut Self, Error>;
+
+    /// Adds P2SH script source for a given output index
+    fn add_output_script(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error>;
+
+    /// Adds P2WSH script source for a given output index
+    fn add_output_wscript(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error>;
+
+    /// Adds information about public key derivation for a given output index
+    /// (for P2PK and P2WPK outputs)
+    fn add_output_derivation(&mut self, index: u32, pubkey: &PublicKey, fingerprint: Fingerprint, path: &DerivationPath) -> Result<&mut Self, Error>;
+}
+
+impl Updater for PartiallySignedTransaction {
+    fn has_sign_info(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn add_input_tx(&mut self, index: u32, tx: &Transaction) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_utxo(&mut self, index: u32, txin: &TxIn) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_script(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_wscript(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_derivation(&mut self, index: u32, pubkey: &PublicKey, fingerprint: Fingerprint, path: &DerivationPath) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_output_script(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_output_wscript(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_output_derivation(&mut self, index: u32, pubkey: &PublicKey, fingerprint: Fingerprint, path: &DerivationPath) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+}
+
+/// Errors which may happen during signer verification
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum SignerVerificationError {
+}
+
+impl fmt::Display for SignerVerificationError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (self as &fmt::Debug).fmt(f)
+    }
+}
+
+/// Implementation of [Role::Signer] role according to
+/// https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#Signer
+pub trait Signer {
+    /// Detects when at least some of the signatures are present
+    fn has_partial_signatures(&self) -> bool;
+
+    /// Verifies that PSBT is ready to be signed, i.e. contains all necessary
+    /// information
+    fn verify(&self) -> Vec<SignerVerificationError>;
+
+    /// Adds signature for a specific input
+    fn add_signature(&mut self, input: u32, signature: Signature) -> Result<&mut Self, Error>;
+}
+
+impl Signer for PartiallySignedTransaction {
```

Comment:
This doesn't sit well with me. I don't understand "how is PartiallySignedTransaction a Signer?". Seems to me that partially signed transaction should be an object/argument of signing and thus ... passed to signer?

I was not paying attention to the matter, so I might be just confused. I'm just skimming through what's going on in the project.

So the goal with the project is to make it hard for people to mess up the workflow?

Should it then be something more like session types? With `PartiallySignedTransaction` remembering its state either at runtime, or in the type itself `PartiallySignedTransaction<State>`?

And then the question is which parts of the role should be swappable and which ones are always the same. 

`verify()` could be a method on  a `PartiallySignedTransaction` if the verification is always the same.

Is it necessary to have a `trait Signer`? Is there anything that actually needs to work on any, arbitrarily provided signer or are these always going to be concrete?

Reply:
"PartiallySignedTransaction accessed from a Signer role perspective"

Why we need roles? Well, we have them in Bitcoin Core, and the purpose of this repo is to follow it. Also, it simplifies passing PSBT around, such that wallets using rust-bitcoin may have frunctions accepting as a parameter some specific role of PSBTs only (`fn(psbt: impl Signer)`).

This is the third iteration, in the previous two we had both of your suggestions: more strirct workflow and PSBT as a parameter to signer object (#456 #457). Here, we have only an ability to add existing signature (to avoid complexity required by async methods) and we will have another `Sync/AsyncSigner` structures for creating signatures with given full PSBT object, which will internally call `add_signature` method from this implementaiton.

Reply:
Is there more than just one way to add an existing signature to PSBT? If not, there is no point of having a `trait`. A method on `PSBT` would do, no? The reason to have a `trait` is to allow extending something with alternative implementations.

The responsibility is role of someone doing something with PSBT, not a state or function of PSBT, no?

Sync and Async do not generalize well in Rust. You might have to do `trait Signer` and `trait AsyncSigner` etc.

Reply:
Just a note: the same stuff we are discussing here applies not only to `Signer` trait, but other role-specific traits defined in this PR.

So, for most of them, they will be implemented not only for `PartiallySignedTransaction`, but also for `Input` and `Output` structs; so yes; we probably need them as traits. Also, I'd like to give wallets ability to limit the scope of what can be done with PSBT object passed as parameter, and by providing traits (even if there is only one structure implementing it) we can do that with generics.